apps/rush-lib/src/logic/ProjectWatcher.ts

```diff
@@ -0,0 +1,197 @@
+
+    this._debounceMilliseconds = debounceMilliseconds;
+    this._rushConfiguration = rushConfiguration;
+    this._selection = selection;
+  }
+
+  /**
+   * Waits for a change to the package-deps of one or more of the selected projects, since the previous invocation.
+   * Will return immediately the first time it is invoked, since no state has been recorded.
+   * If no change is currently present, watches the source tree of all selected projects for file changes.
+   */
+  public async waitForChange(): Promise<IProjectChangeResult> {
+    const initialChangeResult: IProjectChangeResult = this._computeChanged();
+    // Ensure that the new state is recorded so that we don't loop infinitely
+    this._commitChanges(initialChangeResult.state);
+    if (initialChangeResult.changedProjects.size) {
+      return initialChangeResult;
+    }
+
+    const watcher: FSWatcher = new FSWatcher({
+      persistent: true,
+      cwd: Path.convertToSlashes(this._rushConfiguration.rushJsonFolder),
+      followSymlinks: false,
+      ignoreInitial: true,
+      ignored: /(?:^|[\\\/])node_modules/g,
+      disableGlobbing: true,
+      interval: 1000
+    });
+
+    // Only watch for changes in the project folders
+    for (const project of this._selection) {
+      watcher.add(Path.convertToSlashes(project.projectFolder));
+    }
+
+    const watchedResult: IProjectChangeResult = await new Promise(
+      (resolve: (result: IProjectChangeResult) => void, reject: (err: Error) => void) => {
+        let timeout: NodeJS.Timeout | undefined;
+        let terminated: boolean = false;
+
+        const resolveIfChanged = (): void => {
+          timeout = undefined;
+          if (terminated) {
+            return;
+          }
+
+          try {
+            const result: IProjectChangeResult = this._computeChanged();
+
+            // Need an async tick to allow for more file system events to be handled
+            process.nextTick(() => {
+              if (timeout) {
+                // If another file has changed, wait for another pass.
+                return;
+              }
+
+              this._commitChanges(result.state);
+
+              if (result.changedProjects.size) {
+                terminated = true;
+                resolve(result);
+              }
+            });
+          } catch (err) {
+            terminated = true;
+            reject(err);
+          }
+        };
+
+        watcher.on('all', () => {
+          try {
+            if (terminated) {
+              return;
+            }
+
+            // Use a timeout to debounce changes, e.g. bulk copying files into the directory while the watcher is running.
+            if (timeout) {
+              clearTimeout(timeout);
+            }
+
+            timeout = setTimeout(resolveIfChanged, this._debounceMilliseconds);
+          } catch (err) {
+            terminated = true;
+            reject(err);
+          }
+        });
+      }
+    );
+
+    await watcher.close();
```

Comment:
What is the performance impact of repeatedly opening and closing the watcher?

Reply:
As near as I have been able to tell, not much, but then, we have a deliberate 1 second latency before kicking the build since it may take a non-negligible amount of time.